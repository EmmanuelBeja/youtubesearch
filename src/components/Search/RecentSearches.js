import React from 'react'
import { useSelector } from 'react-redux'

import { faCircleNotch } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'

const RecentSearches = ({ searches, handleClick }) => {
  const { loading } = useSelector((state) => state.search)
  const searchItem = searches
  searchItem.reverse()

  const listItems = () => {
    if (loading) {
      return <FontAwesomeIcon icon={faCircleNotch} pulse />
    }

    return searchItem.map((item, index) => (
      <div className="autocomplete__item" key={index} onMouseDown={() => handleClick(item)}>
        {item}
      </div>
    ))
  }

  return (
    <div className={searches.length === 0 ? 'd-none' : 'search__autocomplete'}>
      <div className="autocomplete">
        <div className="autocomplete__info">RECENT SEARCHES</div>
        {listItems()}
      </div>
    </div>
  )
}

export default RecentSearches
