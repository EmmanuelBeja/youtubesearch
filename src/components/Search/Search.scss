.navbar {
  border-top: 5px solid #fe0101;
  border-bottom: 1px solid whitesmoke;
  background-color: white !important;

  .search-section {
    display: flex;
    justify-content: center;
    padding: 0px 50px;
    width: 100%;

    .input-group {
      width: 30%;
      transition: all 0.4s;
      border: 1px solid #ced4da;
      border-radius: 5px;

      &:focus-within {
        width: 35%;

        .autocomplete {
          display: block;
          transition: all 9s;
        }
      }

      @media only screen and (max-width: 576px) {
        width: 98%;

        &:focus-within {
          width: 100%;
        }
      }

      &-prepend {
        background-color: #3f51b514;
        border-radius: 5px 0px 0px 5px !important;
        display: flex;
        justify-content: center;
        align-items: center;
        width: 30px;
      }

      .search__autocomplete > .autocomplete {
        position: absolute;
        left: 0;
        top: 50px;
        right: 0;
        border-radius: 5px;
      }

      .autocomplete {
        display: none;
        background-color: white;
        color: #444;
        box-shadow: 0 1px 2px rgba(0, 0, 0, 0.15);
        font-size: 0.875rem;
        text-align: left;

        &__info {
          color: grey;
          padding: 1rem 2rem 0 2rem;
          font-size: 0.5rem;
        }

        &__item {
          padding: 0.5rem 2rem;
          cursor: pointer;

          &:hover {
            background-color: #eee;
          }
        }
      }

      .form-control {
        border: transparent;
        border-radius: 0px 5px 5px 0px !important;
        background-color: #3f51b514;
        margin-left: 0px !important;

        &:focus {
          outline: none;
          box-shadow: none;
          border-color: transparent;
        }
      }

      &-append {
        button {
          border: none;

          &:hover {
            background-color: white;
            color: black;
          }
        }
      }
    }
  }
}
