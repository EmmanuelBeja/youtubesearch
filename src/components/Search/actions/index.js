import * as types from './actionTypes'
import { searchService } from './service'
import { toast } from 'react-toastify'

const request = (data, type) => ({ type, data })
const success = (data, type) => ({ type, data })
const failure = (error, type) => ({ type, error })

const getSearchResults = (data) => {
  return (dispatch) => {
    dispatch(request(null, types.GET_SEARCH_RESULTS_REQUEST))
    searchService.getSearchResults(data).then(
      (res) => {
        const { items, pageInfo, regionCode } = res
        dispatch(success({ items, pageInfo, regionCode }, types.GET_SEARCH_RESULTS_SUCCESS))
      },
      (error) => {
        toast.error(error?.response?.data?.message)
        dispatch(failure(error.response.data.message, types.GET_SEARCH_RESULTS_FAILED))
      }
    )
  }
}

export const searchActions = {
  getSearchResults,
}
