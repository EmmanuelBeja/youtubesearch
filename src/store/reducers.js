import { combineReducers } from 'redux'
import searchReducer from '../components/Search/reducer'

const allReducers = combineReducers({
  search: searchReducer,
})

const rootReducer = (state, action) => allReducers(state, action)

export default rootReducer
