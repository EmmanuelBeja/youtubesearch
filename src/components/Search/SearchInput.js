import React, { useState, useEffect } from 'react'
import { useSelector, useDispatch } from 'react-redux'

import { InputGroup, InputGroupAddon, Input } from 'reactstrap'
import { useDebounce } from 'use-debounce'
import { faSearch, faCircleNotch } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { toast } from 'react-toastify'
import RecentSearches from './RecentSearches'

import { searchActions } from './actions'

const SearchInput = () => {
  const dispatch = useDispatch()
  const { loading } = useSelector((state) => state.search)

  const [searchInput, setSearchInput] = useState('')
  const [debouncingText, setDebouncingText] = useState('')
  const [debouncedText] = useDebounce(debouncingText, 2000)
  const [inputError, setInputError] = useState(false)
  const [searches, setSearches] = useState([])
  const [showPrevSearches, setShowPrevSearches] = useState(false)

  useEffect(() => {
    debouncedText && handleSearch(debouncedText)
  }, [debouncedText])

  const handleSearch = (input) => {
    if (!input) {
      setInputError(true)
      toast.error('Please type something to search')
      return
    }
    // add searched item to recent searches
    const updatedSearches = searches
    updatedSearches.push(input)

    setSearches(updatedSearches)
    dispatch(searchActions.getSearchResults(input))
    setInputError(false)
  }

  return (
    <InputGroup>
      <InputGroupAddon addonType="prepend">
        {loading ? (
          <FontAwesomeIcon icon={faCircleNotch} pulse />
        ) : (
          <FontAwesomeIcon icon={faSearch} />
        )}
      </InputGroupAddon>
      <Input
        name="searchInput"
        placeholder="Type here..."
        value={searchInput}
        onChange={(e) => {
          setDebouncingText(e.target.value)
          setSearchInput(e.target.value)
        }}
        onFocus={() => setShowPrevSearches(true)}
        onBlur={() => setShowPrevSearches(false)}
        invalid={inputError}
        autoComplete="off"
        disabled={loading}
      />
      {showPrevSearches && <RecentSearches searches={searches} handleClick={handleSearch} />}
    </InputGroup>
  )
}

export default SearchInput
