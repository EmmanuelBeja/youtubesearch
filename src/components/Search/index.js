import React, { useState, useEffect } from 'react'
import { useSelector, useDispatch } from 'react-redux'

import { Navbar, InputGroup, InputGroupAddon, Input } from 'reactstrap'
import { useDebounce } from 'use-debounce'
import { faSearch, faCircleNotch } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { toast } from 'react-toastify'
import RecentSearches from './RecentSearches'

import { searchActions } from './actions'

import './Search.scss'

const Search = () => {
  const dispatch = useDispatch()
  const { loading } = useSelector((state) => state.search)

  const [searchInput, setSearchInput] = useState('')
  const [debouncingText, setDebouncingText] = useState('')
  const [debouncedText] = useDebounce(debouncingText, 2000)
  const [inputError, setInputError] = useState(false)
  const [searches, setSearches] = useState([])

  useEffect(() => {
    debouncedText && handleSearch(debouncedText)
  }, [debouncedText])

  const handleSearch = (input) => {
    if (!input) {
      setInputError(true)
      toast.error('Please type something to search')
      return
    }
    // add searched item to recent searches
    const updatedSearches = searches
    updatedSearches.push(input)

    setSearches(updatedSearches)
    dispatch(searchActions.getSearchResults(input))
    setInputError(false)
  }
  // {{ size: 6, offset: 3 }}
  return (
    <Navbar color="light" light expand="md" className="fixed-top">
      <div className="search-section">
        <InputGroup>
          <InputGroupAddon addonType="prepend">
            {loading ? (
              <FontAwesomeIcon icon={faCircleNotch} pulse />
            ) : (
              <FontAwesomeIcon icon={faSearch} />
            )}
          </InputGroupAddon>
          <Input
            className="form-control"
            name="searchInput"
            placeholder="Type here..."
            value={searchInput}
            onChange={(e) => {
              setDebouncingText(e.target.value)
              setSearchInput(e.target.value)
            }}
            invalid={inputError}
            autoComplete="off"
          />
          <RecentSearches searches={searches} handleClick={handleSearch} />
        </InputGroup>
      </div>
    </Navbar>
  )
}

export default Search

// <Row>
//   <Col
//     lg="12"
//     md="12"
//     sm="12"
//     xs="12"
//     className="search-section shadow"
//   >
//     <InputGroup>
//       <InputGroupAddon addonType="prepend">
//         {loading ? (
//           <FontAwesomeIcon icon={faCircleNotch} pulse />
//         ) : (
//           <FontAwesomeIcon icon={faSearch} />
//         )}
//       </InputGroupAddon>
//       <Input
//         className="form-control"
//         name="searchInput"
//         placeholder="Type here..."
//         value={searchInput}
//         onChange={(e) => {
//           setDebouncingText(e.target.value)
//           setSearchInput(e.target.value)
//         }}
//         invalid={inputError}
//         autoComplete="off"
//       />
//
//       <RecentSearches searches={searches} handleClick={handleSearch} />
//     </InputGroup>
//   </Col>
// </Row>
