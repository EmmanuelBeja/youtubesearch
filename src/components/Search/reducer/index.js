import * as types from '../actions/actionTypes'

const initialState = {
  results: [],
  pageInfo: undefined,
  regionCode: undefined,
  loading: false,
  errors: undefined,
}

const searchReducer = (state = initialState, action) => {
  switch (action.type) {
    case types.GET_SEARCH_RESULTS_REQUEST:
      return {
        ...state,
        results: [],
        loading: true,
      }

    case types.GET_SEARCH_RESULTS_SUCCESS:
      return {
        ...state,
        results: action.data.items,
        pageInfo: action.data.pageInfo,
        regionCode: action.data.regionCode,
        loading: false,
      }

    case types.GET_SEARCH_RESULTS_FAILED:
      return {
        ...state,
        loading: false,
        errors: action.error,
      }

    default:
      return state
  }
}

export default searchReducer
