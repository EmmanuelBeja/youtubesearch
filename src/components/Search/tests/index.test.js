import React from 'react'
import { shallow, mount } from 'enzyme'
import Search from '../'
import SearchInput from '../SearchInput'
import { Provider } from 'react-redux'
import store from '../../../store'

describe('Search', () => {
  const shallowWrapper = shallow(<Search />)
  const mountWrapper = mount(
    <Provider store={store}>
      <Search />
    </Provider>
  )

  it('should render correctly', () => {
    expect(shallowWrapper).toMatchSnapshot()
    expect(shallowWrapper.find('.fixed-top').length).toEqual(1)
  })

  it('should contain SearchInput component', () => {
    expect(mountWrapper.find(<SearchInput />)).toBeTruthy()
  })
})
