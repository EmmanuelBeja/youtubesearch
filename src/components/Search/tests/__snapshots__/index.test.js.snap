// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should render correctly 1`] = `
<Navbar
  className="fixed-top"
  color="light"
  expand="md"
  light={true}
  tag="nav"
>
  <div
    className="search-section"
  >
    <SearchInput />
  </div>
</Navbar>
`;
